Patch: Stacking of series with missing values

benrubson opened the pull request:
Hello,

This PR solves #2912 !

Thank you 👍 

Ben

Commit:
Correct timeStep when there are missing values

Commit:
Correct stack with missing values

Commit:
typo

benrubson commented:
I think [test](https://circleci.com/gh/grafana/grafana/5446) fails because `barWidth` is now calculated over all series of the graph, not only the first one.
If `barWidth` is calculated only using the first series, it can lead to errors if first series has missing values.

Here is an example where series 1 only contains one value every 5 minutes, whereas 2 contains one value every 30 seconds :

![s2](https://cloud.githubusercontent.com/assets/6764151/19110439/3ce1aa38-8afa-11e6-9d6d-7e0375a786f1.png)

The same, calculating `barWidth` over the 2 series :

![s1](https://cloud.githubusercontent.com/assets/6764151/19110441/42051888-8afa-11e6-86b0-22c834bd53a7.png)

I then think the CircleCI test should be updated.

benrubson commented:
@torkelo, could we have this in 3.1.2 please ?
Easy to backport from master as code is the same.
Thank you very much 👍 
Ben

torkelo commented:
not sure, need a lot of testing

Commit:
Update barWidth specs

1/1.5 due to series1 and series2 already declared

benrubson commented:
OK, thank you for your feedback Torkel.
I just have corrected the `graph_specs.ts` file with the new expected result, CircleCI tests now run correctly.
Feel free to ask if I can help with further testing 👍

Commit:
Set tooltip timestamp to the closest point's one

as some series may have missing values

benrubson commented:
Regarding tooltip timestamp commit :

Without :
![s1](https://cloud.githubusercontent.com/assets/6764151/19170602/725edd50-8c19-11e6-84f7-a1ad9d1bd588.png)

With :
![s2](https://cloud.githubusercontent.com/assets/6764151/19170607/77697972-8c19-11e6-8c29-64d26c84c0d7.png)

Commit:
typo

Commit:
typo

torkelo commented on public/app/plugins/panel/graph/graph.ts:
```diff
@@ -273,6 +273,11 @@ module.directive('grafanaGraph', function($rootScope, timeSrv) {
               options.series.bars.barWidth = data[0].stats.timeStep / 1.5;
+              for (let i = 1; i < data.length; i++) {
+                if ((data[i].stats.timeStep / 1.5) < options.series.bars.barWidth) {
+                  options.series.bars.barWidth = data[i].stats.timeStep / 1.5;
```
could you describe this change? Just looking for the series with the smallest timeStep?

benrubson replied:
Yes, this change is looking for the smallest `timeStep`.
If you have several series with different `timeStep`s, you have to choose the smallest one so that bars are not too wide.

torkelo commented:
I am trying to test this PR but I am not seeing any change for how null series are stacked. 

![image](https://cloud.githubusercontent.com/assets/10999/19269181/5803d20a-8fb9-11e6-9c74-fd0b6f05108f.png)

Only one of the series has some null values. You can try this test case yourself as this panel is on a TestData dashboard using the new testdata data source. Just go into plugins > apps > TestData and enable this application. On enabling this app Grafana will import a dashboard named `TestData - Graph Panel Last 1h` which contains this panel.

Commit:
Correct timeStep when there are missing values

more robust version than the previous version, as first series was able to set barWidth to 0.

Commit:
typo

benrubson commented:
@torkelo, regarding your last comment :
_I am trying to test this PR but I am not seeing any change for how null series are stacked._

You're right, I face the same behaviour.
I was focused on series with **missing** values, and did not think about series with **null** values (which is a different case).
I already have a working patch here, I'm finishing it and will push it.

torkelo commented:
@benrubson did you push your latest changes?

benrubson commented:
I plan to do it in a few hours @torkelo 👍

benrubson commented:
See #6287 for a re-edited PR 👍
